### PR TITLE
cpu/msp430_common: Update to inline-able IRQ API

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

- Updated to inline-able IRQ API
- Improved robustness of functions
    - Added memory barrier to prevent the compiler from moving code outside of
      a critical section guarded by irq_disable() ... irq_restore()
- Reduced overhead of `irq_disable()`
    - After clearing the global interrupt enable (GIE) bit, IRQs remain enabled
      for up to one CPU cycle
    - The previous implementation just added a nop to fill that cycle
    - This implementation uses the cycle for masking the return value
- Reduced overhead of `irq_restore()`
    - Now only one CPU cycle is needed
- `irq_disable()`, `irq_restore()`, and `irq_enable()` work now in constant time

#### `tests/irq_disable_restore`

I ran `tests/irq_disable_restore` on a `wsn430-v1_3b` board [in the FIT/IoT-Lab](https://www.iot-lab.info/testbed/experiments/222452):

<details><summary>Console output</summary>

```
s
START
main(): This is RIOT! (Version: 2020.07-devel-1557-g22d44-msp430-irq-inline)
Test for irq_disable() / irq_restore()
======================================

Verifying test works: [SUCCESS]
Test result: [SUCCESS]
```

</details>

#### `tests/tests/bench_mutex_pingpong`

I ran `tests/tests/bench_mutex_pingpong` on the same hardware in the [FIT/IoT-Lab](https://www.iot-lab.info/testbed/experiments/222454)

<details><summary>Console output with `master`</summary>

```
r
READY
s
START
main(): This is RIOT! (Version: 2020.07-devel-1557-g22d44)
main starting
{ "result" : 9222 }
```

</details>

<details><summary>Console output with this PR</summary>

```r
READY
s
START
main(): This is RIOT! (Version: 2020.07-devel-1558-gc5c83-msp430-irq-inline)
main starting
{ "result" : 9910 }
```

</details>

So ~7.5% speedup.

### Issues/PRs references

Follow up of: https://github.com/RIOT-OS/RIOT/pull/13999

Tracked in: https://github.com/RIOT-OS/RIOT/issues/14356

### CODE REVIEW

**Comment by fjmolinas** [comment]
I'll test on `z1`

**Comment by aabadie** [comment]
> on a wsn430-v1_3b board in the FIT/IoT-Lab:

That might be one of the last experiment on these devices: they are going to be removed during the summer :(

File: cpu/msp430_common/include/irq_arch.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+#define IRQ_ARCH_H
+
+#include "irq.h"
+#include "cpu.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+extern volatile int __irq_is_in;
```
Comment (fjmolinas) [inline cpu/msp430_common/include/irq_arch.h:33]: why is volatile the best choice here?

**Comment by kaspar030** [comment]
I think this one will conflict a lot with https://github.com/RIOT-OS/RIOT/pull/12457. That'll have to wait until I'm back from vacation, but IMO it would make sense to base this PR on that.

**Comment by fjmolinas** [comment]
> I think this one will conflict a lot with #12457. That'll have to wait until I'm back from vacation, but IMO it would make sense to base this PR on that.

Do the instructions for `isr` handling change? Where do you see the issues exactly?

File: cpu/msp430_common/include/irq_arch.h (hunk 1 above)
Comment (maribu) [reply cpu/msp430_common/include/irq_arch.h:33]: IMO it should *not* be `volatile`. But I didn't want to change this, as this not directly related to the API transition.

**Comment by maribu** [comment]
> I think this one will conflict a lot with #12457.

I just rebased #12457 on top of this PR. These were the conflicts:

- `cpu/msp430_common/include/stdatomic.h`
- `boards/chronos/doc.txt`
- `boards/chronos/Kconfig`
- `tests/gnrc_ipv6_nib/Makefile.ci`
- `sys/net/gnrc/transport_layer/tcp/gnrc_tcp_pkt.c`

None of them were caused by this PR. (And not was terrible difficult to fix; the first three files have just be removed, the `Makefile.ci` was trivial, and the `gnrc_tcp_pkt.c` was just the replacement of `abs()` with `labs()`, which can easily be redone on top of the new code.)

**Comment by kaspar030** [comment]
> I just rebased #12457 on top of this PR.

thanks for checking!

**Comment by maribu** [comment]
Anyway, soft feature freeze in effect since 20 minutes. So I guess this will have to wait until next release :-/

**Comment by fjmolinas** [comment]
Results:

- https://ci.inria.fr/ci-riot-tribe/job/build-pipeline.jk/lastSuccessfulBuild/artifact/failuresummary.md/*view*/
- details https://ci.inria.fr/ci-riot-tribe/job/build-pipeline.jk/lastSuccessfulBuild/artifact/z1/

Considering https://github.com/RIOT-OS/RIOT/issues/13267 I think failures are OK!

**Comment by fjmolinas** [comment]
I'm not familiar with the msp430 instruction set, I will need to read on it a bit. But since feautre freeze is in, there is no real urgency :)

**Comment by fjmolinas** [comment]
> I've looked at the assembler code and from my understanding looks OK, I had already tested before. With your change there is no usage of __enable_irq() and __disable_irq() is only used in thread_yield_higher, I think those private functions should be removed from cpu/msp430_common/include/cpu.h and the occurrence of __disable_irq replaced by irq_disable().

If you agree that those function should be removed, then please trigger murdock afterwards @maribu!

**Comment by fjmolinas** [comment]
@maribu Maybe remove those functions in a distinct commit instead of a fixup? We can wait for murdock to finish for that to see if other things pop up.

**Comment by maribu** [comment]
Some Murdock workers seem to have issues with the network file system right now :-/ Maybe someone from the CI team can take a look?

**Comment by fjmolinas** [comment]
@riot-ci @RIOTCI @aabadie @kaspar030 @cgundogan any ideas what is going on with murdock?

**Comment by cgundogan** [comment]
>@riot-ci @RIOTCI @aabadie @kaspar030 @cgundogan any ideas what is going on with murdock?

there was a test server running out of memory. That is removed now from the build queue. I retriggered the affected PRs.

**Comment by maribu** [comment]
@cgundogan: Thanks! :-)

All green now :-)

**Comment by maribu** [comment]
Thanks :-)

**Pull request merged**